### ISSUE TITLE
How could Docz use Gatsby as bundler?

### ISSUE THREAD

**Issue by pedronauck** [doczjs/docz#810]
One of our main bottlenecks on Docz is the fact that we need to handle with a lot of dependencies in order to have our core running with a Javascript bundler. This doesn’t make much sense, since our main focus should be to provide out of the blue a full environment with basic components and, with that, create a very simple way to write documentation.

Have to deal with a bundler like Webpack on our core is a pretty complex and verbose task that creates a lot of issue for us. Because it has so many configurations and dependencies that we need to handle - and I'm not even getting into the issue of SSR/Static build that we don’t have natively - it's cause of a bunch of incompatibilities and we need to put a lot of effort trying to solve this problem instead of creating new features and improvements.

I know that this problem is something that we always will have since we want to give for our users a zero-config and easy environment. But I think that has a lot better ways to handle with that, like using some tool as a bundler instead of deal with our own. Thinking about that, I really think that use Gatsby is the best choice for the task because:

- [x] Gatsby is fast and has a so powerful dev experience
- [x] It has a pretty big and concrete ecosystem around
- [x] Docz as a product will simplify potentially in terms of structure
- [x] Our issues around bundler will decrease a lot

## Docz structure and algorithm

In order to provide a cleaner history, I’ll give a short explanation of how our structure and algorithms work and what each package does inside our monorepo.

First of all, we have this to illustrate both algorithms of `docz dev` and `docz build` commands:

![](https://cdn-std.dprcdn.net/files/acc_649651/tgKkIJ)
![](https://cdn-std.dprcdn.net/files/acc_649651/Q8uHI6)

Basically, these images explain these codes:
- https://github.com/pedronauck/docz/blob/master/core/docz-core/src/commands/dev.ts
- https://github.com/pedronauck/docz/blob/master/core/docz-core/src/commands/build.ts

So, about our package structure, if you look on it at the first time, you will think that is very complex in terms of packages, because of a lot of them are isolated, but it’s not so hard to figure out how things work.

In the short term, basically, we have these packages:

- **docz:** cli, core components and hooks
- **docz-core:** bundler, data handling and parsers.
- **docz-theme-default:** just a bunch of React components that use `docz` built-ins components and `docz-core` to create an application (in our case, a documentation app).
- **docz-utils:** a bunch of node methods used inside our packages
- **gatsby-theme-docz:** our Gatsby theme
- **rehype-docz:** rehype plugin that handles with some AST specific things for docz
- **remark-docz:** same of above but for remark

And in the long term:

### `docz`

* Main and top-level package.
* Scripts that are running on the browser (like components and hooks) belongs to this package
* Built-ins components are built here, most specifically on [this folder](https://github.com/pedronauck/docz/tree/master/packages/docz/src/components) 
* This package shouldn’t have any component style, just boilerplate and logic!
* The CLI's setup lives here, but all functions came from `docz-core`. This is just a “trick” to enables the user to don’t need to install two packages.

---

### `docz-core`

* This is the core of docz. All bundler configuration, build algorithms, data handling, server processes and parses belongs to here.
* We have few classes that are the base of our core to deal with our some algorithms:

#### `DataServer`

One of the main features here is about the data that is parsed using AST and generated inside the `.docz/` folder as a JSON database to be consumed by the user inside the built-ins components. All this data is handled by the [`DataServer`](https://github.com/pedronauck/docz/blob/master/core/docz-core/src/lib/DataServer.ts) and it's used inside Gatsby theme as well in order to create the source nodes.

```js
const dataServer = new DataServer()

// this will register which "states" we need
if (config.propsParser) dataServer.register([states.props(config)])
dataServer.register([states.config(config), states.entries(entries, config)])

// this will generate our first instance of the data
await dataServer.start()

// and watching the states change we can do something like a socket message
socket.onConnection((_, emit) => {
  dataServer.onStateChange(action => {
    emit(action.type, action.payload)
  })
})
```

#### `Entries` and `Entry`

Another important feature applied to our core is the ability to find `.mdx` files anywhere, parse them and extract information from it. This is possible because of the [`Entries`](https://github.com/pedronauck/docz/blob/master/core/docz-core/src/lib/Entries.ts) and the [`Entry`](https://github.com/pedronauck/docz/blob/master/core/docz-core/src/lib/Entry.ts) classes.

```js
// create an instance of Entries using the parsed config object
// this `config` object is the result of arguments and doczrc.js configuration merged
const entries = new Entries(config)

// will find, parse and retrieve a list of entries with all data of each filepath
await entries.get()
```

#### `Bundler`
Last but not least, it's the bundler stuff living inside the `docz-core`. All bundler configuration lives inside the [`bundler/`](https://github.com/pedronauck/docz/tree/master/core/docz-core/src/bundler) folder, here we have configurations for loaders, plugins, `webpack-dev-server`, minifiers, etc, all separated by files and each one dealing with your own concern. So, inside this folder have just the configuration, the responsibility to manage this config is of the `Bundler` class:

```ts
export interface BundlerConstructor<Config> {
  args: Args
  config: ConfigFn<Config>
  server: ServerFn<Config>
  build: BuildFn<Config>
}

export const bundler = (args: Args, env: Env): Bundler<CFG> =>
  new Bundler({
    // this `args` variable is our parsed config from arguments and doczrc
    args,
    // function used for build with the bundler
    build,
    // function that will provide a default configuration for the bundler
    config: createConfig(args, env),
    // function that will run the developer server
    server: server(args),
  })
```

When I decided to create this `Bundler` class I was thinking that in the future I could have another's bundlers for Docz and not just Webpack, so create some kinda a high-level API to deal with that would help me to change or create a new one easily. Exactly because that, use this class is something trivial like you can see inside the `commands` function used on CLI:

```js
import { bundler as webpack } from './bundler'

function dev() {
  // first create the instance of the bundler
  const bundler = webpack(config, env)
  // then, create the configuration based on the environment (dev and prod)
  const bundlerConfig = await bundler.mountConfig(env)
  // then, create the server instance and start it
  const app = await bundler.createApp(bundlerConfig)
  const server = await app.start()
}

function build() {
  // like in the dev function, create the instance and configuration
  const bundler = webpack(config, env)
  const bundlerConfig = await bundler.mountConfig(env)

  // but for production, just run the build method instead of starting a server
  await bundler.build(bundlerConfig)
}
```

For now, that's the only necessary information that you need to know in order to change our current bundler (webpack) and use Gatsby instead of it.

## Using Gatsby programmatically

Docz came up with some project goals in mind about dev experience:

- [x] Zero-config, simple and flexible to use
- [x] Run in parallel with any type of project (create-react-app, nextjs, gatsby, etc)
- [x] Don't need to have strict folders and files structures in order to run

So, our main challenge in order to get Docz up and running without big changes for the users and without hurt any of these project goals is to be able to consume Gatsby with a programmatically API like we're doing with Webpack. With that in mind, we could change just the bundle stuff inside our core and basically use the same structure that we have today since we already made a lot of changes in order to fit our structure to create our Gatsby theme.

So, some abstract and super high-level code for devserver could be something like that:

```ts
// docz-core/src/bundlers/devserver.ts
import { gatsby, Config } from 'gatsby-node'

import { Config as Args } from '../config/argv'
import { ServerHooks as Hooks } from '../lib/Bundler'

export const server = (args: Args) => async (config: Config, hooks: Hooks) => ({
  start: async () => {
    // instantiate gatsby app
    const app = gatsby({
      pagesDir: './src/pages',
      env: 'development',
      browser: {},
      config: {},
      node: {},
      ssr: {},
      /* some another config */
    })

    // like in webpack-dev-server this method can return the server instance
    const devServer = app.devServer()
    devServer.listen(config.port, args.host, err => {
      if (err) return app.throw(err)
      hooks.onServerListening(devServer)
    })

    return devServer
  },
})
```

This is quite similar (almost the same) that we have today and could fit very well.
And for build we could have something like that:

```ts
import { gatsby, Config } from 'gatsby-node'

export const build = async (config: Config, dist: string, publicDir: string) => {
  const app = gatsby({
    pagesDir: './src/pages',
    env: 'production',
    browser: {},
    config: {},
    node: {},
    ssr: {},
    /* some another config */
  })

  await app.build(dist)
}
```

We could simplify this a lot like creating the app instance just in one place and passing it as an argument, but this is a **so much abstract** code just to illustrate how we could have Gatsby working in a very programmatically way.

## Potential issues

- [ ] Webpack incompatibilities problems when running in parallel with `create-react-app` and some other build frameworks.
- [ ] The `public` folder generated by default on Gatsby could be a problem, but I think that in this case we can put all Gatsby generated stuff in a folder like `.gatsby` like we do with `.docz`
- [ ] The default `src` folder can be a problem as well to running parallel with other projects, will be nice to have an option for that in the api config, but this can be a huge messy since a lot of Gatsby plugins/transformer/themes assume their algorithms based on this folder 😕
- [ ] ...

## How could we start?

I think that the first thing is to study how we could implement Gatsby as a programmatically API and then make the necessary modifications inside Docz.

What do you think about that? @KyleAMathews @ChristopherBiscardi @johno @jnxblk

**Comment by ivan-dalmet** [comment]
I don't know much about Gatsby so I will ask question as a user of Docz.

>  Webpack incompatibilities problems when running in parallel with create-react-app and some other build frameworks.

Is it different from now? We run through a lot of conflicts with the way Docz is bundled today. So the only problem might be that we don't have the power to do immediate changes if Docz use Gatsby bundler. But I don't think this is a major issue if Docz works out of the box with create-react-app at least.

> The `public` folder generated by default on Gatsby could be a problem, but I think that in this case we can put all Gatsby generated stuff in a folder like `.gatsby` like we do with `.docz`

Yes, using a `.gatsby` or a `.docz/gatsby` folder is necessary.

> The default `src` folder can be a problem as well to running parallel with other projects, will be nice to have an option for that in the api config, but this can be a huge messy since a lot of Gatsby plugins/transformer/themes assume their algorithms based on this folder 😕

I'm not very familiar with Gatsby and its ecosystem. But if there a config in gatsby to change the `src` folder, the plugins/transformer/themes should use the config? If not, it's not really the Docz problem I think (but again I'm not a Gatsby user, so I don't know if it's a major issue or not).

**Comment by nitin-** [comment]
How to skip checking for ssr
